go/consensus/tendermint/apps/roothash/roothash.go

```diff
@@ -560,6 +565,59 @@ func (app *rootHashApplication) tryFinalizeExecutorCommits(
 
 		ec := commit.ToDDResult().(*commitment.ExecutorCommitment)
 
+		// Update the incoming message queue by removing processed messages. Do one final check to
+		// make sure that the processed messages actually correspond to the provided hash.
+		state := roothashState.NewMutableState(ctx.State())
+		if ec.Header.InMessagesCount > 0 {
+			var meta *message.IncomingMessageQueueMeta
+			meta, err = state.IncomingMessageQueueMeta(ctx, rtState.Runtime.ID)
+			if err != nil {
+				return fmt.Errorf("failed to fetch incoming message queue metadata: %w", err)
+			}
+			var msgs []*message.IncomingMessage
+			msgs, err = state.IncomingMessageQueue(ctx, rtState.Runtime.ID, 0, ec.Header.InMessagesCount)
+			if err != nil {
+				return fmt.Errorf("failed to fetch incoming message queue: %w", err)
+			}
+			if inMsgsHash := message.InMessagesHash(msgs); !ec.Header.InMessagesHash.Equal(&inMsgsHash) {
+				ctx.Logger().Debug("finalized round contained invalid incoming message hash, failing instead",
+					"in_msgs_hash", inMsgsHash,
+					"ec_in_msgs_hash", *ec.Header.InMessagesHash,
+				)
+				// Make the round fail.
+				err = fmt.Errorf("finalized round contained invalid incoming message hash")
+				// TODO: All nodes contributing to this round should be penalized.
```

Comment:
Is there an issue opened for this?

Reply:
No, there is an internal task about adding slashing/suspension that I plan to work on next and I thought that should be part of it. Will open an issue as well.